### ISSUE TITLE
📝 Formatter randomly hangs

### ISSUE THREAD

**Issue by noherczeg** [biomejs/biome#3049]
### Environment information

```bash
CLI:
  Version:                      1.7.1
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v20.13.1"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "npm/10.5.2"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              true
  Organize imports disabled:    false
  VCS disabled:                 true

Formatter:
  Format with errors:           false
  Indent style:                 Space
  Indent size:                  0
  Indent width:                 2
  Line ending:                  Lf
  Line width:                   120
  Attribute position:           Auto
  Ignore:                       []
  Include:                      []

JavaScript Formatter:
  Enabled:                      false
  JSX quote style:              Double
  Quote properties:             AsNeeded
  Trailing comma:               All
  Semicolons:                   Always
  Arrow parentheses:            Always
  Bracket spacing:              false
  Bracket same line:            false
  Quote style:                  Single
  Indent style:                 unset
  Indent size:                  unset
  Indent width:                 unset
  Line ending:                  unset
  Line width:                   unset
  Attribute position:           Auto

JSON Formatter:
  Enabled:                      true
  Indent style:                 unset
  Indent width:                 unset
  Indent size:                  unset
  Line ending:                  unset
  Line width:                   unset
  Trailing Commas:              unset

Workspace:
  Open Documents:               0
```

### Configuration

```JSON
{
  "$schema": "./node_modules/@biomejs/biome/configuration_schema.json",
  "organizeImports": {
    "enabled": true
  },
  "formatter": {
    "enabled": true,
    "formatWithErrors": false,
    "indentStyle": "space",
    "indentWidth": 2,
    "lineWidth": 120,
    "ignore": []
  },
  "javascript": {
    "formatter": {
      "quoteStyle": "single",
      "trailingComma": "all"
    }
  },
  "linter": {
      "enabled": false
  }
}
```

### Playground link

https://biomejs.dev/playground/?code=LwAvACAAUwBlAGUAIABoAHQAdABwAHMAOgAvAC8AZwBpAHQAaAB1AGIALgBjAG8AbQAvAG4AbwBoAGUAcgBjAHoAZQBnAC8AYgBpAG8AbQBlAC0AYQBnAHQALQB0AGUAcwB0ACAAZgBvAHIAIABhAGMAdAB1AGEAbAAgAGMAbwBkAGUAYgBhAHMAZQAuAC4ALgA%3D

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by noherczeg** [comment]
There is no point checking the playground, we can only reproduce our issue with a sufficiently large codebase.

We are having issues with the formatter hanging. This has been going on for months now, but only now were we able to provide a project where it occurs around 2-3 times per 10 runs.

The formatter hangs randomly, or not at all:

- in the middle of formatting
- at the end after printing how many files have been formatted

Please see the example codebase:https://github.com/noherczeg/biome-agt-test

Keep re-running the `pnpm run format` command until the issue occurs. If at the end of the run it doesn't show up, `git reset --hard` the repo, and run it again until it does.

This is happening on Both M1 MacOS, Ubuntu, and Fedora x64 as well.

The larger the codebase, the more likely it hangs.

**Comment by ah-yu** [comment]
Looks like it is a deadlock issue and I believe it was fixed in https://github.com/biomejs/biome/pull/2753. Can you try the nightly version of biome and see if the issue persists

**Comment by ematipico** [comment]
It's a deadlock issue indeed. Sorry you're experiencing this @noherczeg 

Can you follow @ah-yu suggestions, and try our latest nightly, and report to us?

**Comment by noherczeg** [comment]
@ah-yu , @ematipico tested it with 30 consecutive runs, and the issue didn't occur any more. I think the fix works. Thank you all!

**Comment by noherczeg** [comment]
When can we expect a full release?

**Comment by Conaclos** [comment]
@noherczeg We just released Biome v1.8 :)
